### PR TITLE
[CUDAGraph] Add getter for cuda graph exec

### PR DESCRIPTION
**Pull request by galv**
This is far simpler than #155164 since we never destroy the cudaGraphExec_t.

The request comes from TRT-LLM specifically. The motivation is that some power users would like to mutate specific kernel parameters via APIs like `cudaGraphExec*SetParams` after a cuda graph has been instantiated. For example, a common request has been to be able to change the sequence length of attention kernels, after having captured a graph for the largest possible sequence length. It turns out that the host overhead you eliminate via cuda graphs in LLM inference ends up causing an increase in computation time when you size your kernels to the maximum possible sequence length (which I believe is done in both TRT-LLM and vLLM). Attention is the most problematic kernel because its computation time is quadratic in the sequence length, rather than linear.

This can work if your attention kernel can work for arbitrary shapes (this is not the case for all attention implementations! Many of them specialize with templates), and you have a persistent kernel that allocates only as many blocks as you have SM's (so you don't have to figure out how many blocks to allocate for a specific sequence length). Using a conditional SWITCH node is a better generic approach to this problem, but that requires more infrastructure work.

Note that this requires knowledge of the exact location of the value in your kernel's parameter buffer to mutate. It won't work with arbitrary stream capture code whose kernels you don't know before hand. So I expect this code path to be rarely used.

Testing:

```
pytest -s -k raw_graph_exec test/test_cuda.py
```

cc @mcarilli @ezyang @eellison @penguinwu @BoyuanFeng

### CODE REVIEW

File: test/test_cuda.py [hunk 1]
```diff
@@ -3639,6 +3639,26 @@ def test_cuda_graph_raw_graph_keep_graph_false(self):
...
+        not TEST_CUDA_GRAPH or not TEST_CUDA_PYTHON_BINDINGS,
+        "CUDA >= 11.0 or ROCM >= 5.3 required for graphs, cuda-bindings must be installed",
+    )
+    def test_cuda_graph_raw_graph_exec_keep_graph_false(self):
+        import cuda.bindings.runtime as cudart
+
+        graph = torch.cuda.CUDAGraph(keep_graph=False)
+        x = torch.zeros([2000], device="cuda")
+        y = torch.ones([2000], device="cuda")
+        with torch.cuda.graph(graph, capture_error_mode="relaxed"):
```
Comment (ngimel) [inline test/test_cuda.py]: ooc why relaxed mode here?

File: test/test_cuda.py [hunk 2]
```diff
@@ -3666,6 +3686,33 @@ def test_cuda_graph_raw_graph(self):
 
         graph.replay()
 
+    @unittest.skipIf(
+        not TEST_CUDA_GRAPH or not TEST_CUDA_PYTHON_BINDINGS,
+        "CUDA >= 11.0 or ROCM >= 5.3 required for graphs, cuda-bindings must be installed",
+    )
+    def test_cuda_graph_raw_graph_exec(self):
```
Comment (ngimel) [inline test/test_cuda.py]: seems like this test can be parametrized on keep_graph, there's very little difference

**Comment by BoyuanFeng** [comment]
Just curious.

1. Could you provide a pytorch eager example on how to use this feature with attention ops? Many users would be interested.
2. For the variable length attention use case. FWIW, users (e.g., vllm) usually capture CUDAGraphs for each batch size as multiples of 8 and pad inputs to nearest batch size. What's the advantage of the proposed feature?
3. Any latency estimation of updating this config during runtime?

File: test/test_cuda.py (hunk 1 above)
Comment (galv) [reply test/test_cuda.py]: I haven't done any "warmup" in this case, and it's possible to run this test before any other test. On the chance that torch.ops.aten.add does some kind of "graph unsafe" operation (I don't remember if this is the case to be honest, but you can see I did this with the keep_graph tests), doing stream capture in "relaxed" mode prevents these unsafe operations from causing an error.

File: test/test_cuda.py (hunk 2 above)
Comment (galv) [reply test/test_cuda.py]: Oh yes this is right. Will do.

File: test/test_cuda.py (hunk 1 above)
Comment (galv) [reply test/test_cuda.py]: if you want me to use warmup first and then do thread local mode, that is fine with me. Just say so. I don't really like to sweat details like this.

**Review by ngimel — APPROVED** [review]

**Review by BoyuanFeng — APPROVED** [review]

**Review by eellison — APPROVED** [review]

**Review by eqy — APPROVED** [review]

**Comment by ngimel** [comment]
@pytorch merge -i
